### ISSUE TITLE
False Negative: Fall-through variants tied to comment placement and case organization are missed.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6530]
# False Negative: Fall-through variants tied to comment placement and case organization are missed.

Version
7.22.0

## Checker
- Checker id: `category/java/errorprone.xml/ImplicitSwitchFallThrough`
- Checker description: Detects switch branches that implicitly fall through to the next branch without a comment indicating the fall-through is intentional.

## Description of the false negative
`diff/output` marks 2 false negatives for `category/java/errorprone.xml/ImplicitSwitchFallThrough`. Across these samples, the benchmark is still exercising the same underlying bug pattern, but the implementation appears to lose track of it once the source deviates from one canonical shape. The recurring themes are switch control flow.

### `PosCase2.java`
Source: `output/errorprone/FallThrough/src/main/java/scensct/core/pos/PosCase2.java`

The example is built around a `switch` whose control-flow shape is slightly unusual because comments, `break`s, or `default` handling change how a human would read the fall-through. The case layout is still suspicious; the only variation is where the comment or `break` appears. That is also how the benchmark classifies the sample: Switch case with break but a fall through comment should be flagged as impossible fall through.

```java
// Switch case with break but a fall through comment should be flagged as impossible fall through.
package scensct.core.pos;

public class PosCase2 {
    public void testImpossibleFallThrough() {
        int x = 1;
        switch (x) {
            case 1:
                System.out.println("Case 1");
                break; // Case cannot complete normally due to break
                // fall through (comment present but fall-through is impossible)
            case 2: // Immediately following case with a fall through comment // [REPORTED LINE]
                System.out.println("Case 2");
                break;
            default:
                break;
        }
    }
}
```

### `PosCase2.java`
Source: `output/sonarqube/SwitchCaseWithoutBreakCheck/src/main/java/scensct/core/pos/PosCase2.java`

The example is built around a `switch` whose case-group layout is the whole point of the test. The risky pattern is still present in the code, so the absence of a report points to a matching gap rather than a benign variant. That is also how the benchmark classifies the sample: Switch case group CG1 falls through to CG2 with comments in CG2 body not on same line as CG1's first token should be flagged as unintentional fall-through.

```java
// Switch case group CG1 falls through to CG2 with comments in CG2 body not on same line as CG1's first token should be flagged as unintentional fall-through.
package scensct.core.pos;

public class PosCase2 {
    public static void main(String[] args) {
        int x = 1;
        switch (x) {
            case 1:
                System.out.println("CG1");
                // Falls through, but comment is not on same line as 'case 1:'
            case 2:
                // This comment is inside CG2 body, not on same line as 'case 1:'
                System.out.println("CG2");
                break;
        }
    }
}
```

## Cause analysis
The rule is documented as follows: Detects switch branches that implicitly fall through to the next branch without a comment indicating the fall-through is intentional.

In this set the problem shows up around switch control flow. The missed cases suggest that the implementation is narrower than the rule description and too dependent on one preferred AST shape.

Once the same behavior is expressed through a helper, an overload, a modifier such as `abstract` or `native`, a different comment position, or a slightly rearranged control-flow structure, the report disappears even though the benchmark still treats the case as in-scope.

For `category/java/errorprone.xml/ImplicitSwitchFallThrough`, the practical improvement would be to generalize the match so it follows the underlying behavior rather than one exact spelling of it.

## References
None known.

**Comment by oowekyala** [comment]
Reporting fall through comments in cases that don't actually fall through could be considered to be added for the rule but as it stands it is not part of the current scope of the rule. Will close this issue as it is difficult to read, unfocused and unactionable without rewriting it.
